vault/core.go

```diff
@@ -738,19 +737,57 @@ func (c *Core) checkToken(
 		return nil, nil, ErrInternalError
 	}
 
+	return acl, te, nil
+}
+
+func (c *Core) checkToken(req *logical.Request) (*logical.Auth, *TokenEntry, error) {
+	defer metrics.MeasureSince([]string{"core", "check_token"}, time.Now())
+
+	acl, te, err := c.fetchACLandTokenEntry(req)
+	if err != nil {
+		return nil, nil, err
+	}
+
 	// Check if this is a root protected path
-	if c.router.RootPath(path) && !acl.RootPrivilege(path) {
-		return nil, nil, logical.ErrPermissionDenied
+	rootPath := c.router.RootPath(req.Path)
+
+	resourceExists := new(bool)
+	if req.Operation == logical.CreateOperation || req.Operation == logical.UpdateOperation {
```

Comment:
I would add a comment explaining that we get a WriteOperation and then specialize it as either Create or Update using the existence check